plonk-hashing/src/poseidon/poseidon.rs

```diff
@@ -0,0 +1,584 @@
+    fn linear_combination(
+        c: &mut StandardComposer<F, P>,
+        state: &[Self::Field; WIDTH],
+        coeff: impl IntoIterator<Item = Self::ParameterField>,
+    ) -> Self::Field {
+        let coeffs = coeff.into_iter().collect::<Vec<_>>();
+        let mut remaining = WIDTH;
+        let mut index = 0;
+        let mut result: Self::Field;
+        // the first time you have no accumulated result yet, so you can take 3
+        // inputs
+        if remaining < 3 {
+            // this is unlikely, WIDTH is usually at least 3
+            result = c.arithmetic_gate(|g| {
+                g.witness(state[0], state[1], None)
+                    .add(coeffs[0], coeffs[1])
+            });
+            remaining -= 2;
+        } else {
+            result = c.arithmetic_gate(|g| {
+                g.witness(state[index], state[index + 1], None)
+                    .add(coeffs[index], coeffs[index + 1])
+                    .fan_in_3(coeffs[index + 2], state[index + 2])
+            });
+            index += 3;
+            remaining -= 3;
+        }
+
+        // Now you have an accumulated result to carry, so can only take 2
+        // inputs at a time
+        while remaining > 0 {
+            if remaining < 2 {
+                // Accumulate remaining one
+                result = c.arithmetic_gate(|g| {
+                    g.witness(state[index], result, None)
+                        .add(coeffs[index], Self::ParameterField::one())
+                });
+                remaining -= 1;
+            } else {
+                // Accumulate next two
+                result = c.arithmetic_gate(|g| {
+                    g.witness(state[index], state[index + 1], None)
+                        .add(coeffs[index], coeffs[index + 1])
+                        .fan_in_3(Self::ParameterField::one(), result)
+                });
+                index += 2;
+                remaining -= 2;
+            }
+        }
+        result
+    }
+
+    #[cfg(not(feature = "no-optimize"))]
+    fn product_mds_with_sparse_matrix(
+        c: &mut StandardComposer<F, P>,
+        state: &mut [Self::Field; WIDTH],
+        matrix: &SparseMatrix<Self::ParameterField>,
+    ) {
+        let mut result = Self::zeros::<WIDTH>(c);
+
+        result[0] =
+            Self::linear_combination(c, state, matrix.w_hat.iter().cloned());
+        for (j, val) in result.iter_mut().enumerate().skip(1) {
+            // for each j, result[j] = state[j] + state[0] * v_rest[j-1]
+            *val = c.arithmetic_gate(|g| {
+                g.witness(state[0], state[j], None)
+                    .add(matrix.v_rest[j - 1], F::one())
+            });
+        }
+        *state = result;
+    }
+}
+
+trait HashFunction<const WIDTH: usize, COM = ()> {
+    type Input;
+    type Output;
+
+    // The input ought to have size ARITY, but const generic expressions aren't
+    // allowed yet
+    fn hash(&self, input: &[Self::Input], compiler: &mut COM) -> Self::Output;
```

Comment:
We can use const-generics. See this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=da23a1a6286e7681a760f7d1bdfa55a7.

Reply:
But can we use const-generic _expressions_?  The issue I was referring to is that as you said above, `HashFunction` trait should have the generic `ARITY` but within `PoseidonSpec` we need to arrays whose size is `ARITY + 1`.  That produces "error: generic parameters may not be used in const operations" with the suggestion to "use `#![feature(generic_const_exprs)]` to allow generic const expressions"

Reply:
The notion of width is internal to Poseidon, so the way it's represented in memory is an implementation detail, not something that is exposed to the interface. Using arrays would be nice but since the feature is unstable, the other options are to use a `Vec` for the state vector (of length `ARITY + 1`) or to separate the state into two parts like this:

```rust
type State = (F, [F; ARITY]);
```

In either case, the input to the hash function is of the form `[F; ARITY]`, the rest of the state is not part of the interface.

NB: In my implementation in `manta-rs` I use `Vec` for simplicity but for the production circuit I'm going to try and use `State` as above.